### PR TITLE
Core parser refactor

### PR DESCRIPTION
**Pull request by jonsequitur**
The previous implementation of the core parser coupled parsing and syntax concerns with binding concerns. There are a few issues I'm planning to address with this refactor.

* `ArgumentResult` was not a `SymbolResult` while `CommandResult` and `OptionResult` were. `SymbolResult.Children` did not contain the `ArgumentResult` objects. This is confusing.
* `SymbolResult` coupled binding, parsing, and validation in ways that were hard to follow.
* Changing the parsing rules independently of the binding rules was therefore difficult.

I'm separating the parsing functionality with the goal of slimming down the `*Result` types. New parser types are in the `System.CommandLine.Parsing` namespace.

### CODE REVIEW

File: src/System.CommandLine/StringExtensions.cs [hunk 1]
```diff
@@ -210,6 +200,74 @@ void ReadResponseFile(string filePath, int i)
...
+            }
+            else
+            {
+                args = new[]
+                       {
+                           commandName
+                       }.Concat(args).ToList();
+            }
+
+            return args.ToList();
```
Comment (Keboo) [inline src/System.CommandLine/StringExtensions.cs]: Both branches of the above if statement call `.ToList()`. You can likely remove them in favor of just this one.

File: src/System.CommandLine/SymbolExtensions.cs [hunk 2]
```diff
@@ -15,14 +16,14 @@ internal static class SymbolExtensions
...
-             !string.IsNullOrWhiteSpace(symbol.Description) ||
+            (!String.IsNullOrWhiteSpace(symbol.Name) ||
+             !String.IsNullOrWhiteSpace(symbol.Description) ||
              symbol.Arguments().Any(a => a.ShouldShowHelp()));
 
         internal static bool ShouldShowHelp(
             this IArgument argument) =>
             argument != null &&
-            !string.IsNullOrWhiteSpace(argument.Name) &&
+            !String.IsNullOrWhiteSpace(argument.Name) &&
```
Comment (Keboo) [inline src/System.CommandLine/SymbolExtensions.cs]: Is there a reason for converting away from the keyword?

File: src/System.CommandLine/SymbolResult.cs [hunk 3]
```diff
@@ -25,293 +28,220 @@ public abstract class SymbolResult
...
+                return ArgumentResults.SingleOrDefault() ??
+                       ArgumentResult.None(argument);
+            }
+        }
+
+        internal ArgumentResultSet ArgumentResults
+        {
+            get
+            {
+                // FIX: (ArgumentResults)
```
Comment (Keboo) [inline src/System.CommandLine/SymbolResult.cs]: Not sure I understand this comment

File: src/System.CommandLine/SymbolResult.cs [hunk 4]
```diff
@@ -25,293 +28,220 @@ public abstract class SymbolResult
...
+            return _defaultArgumentValues.GetOrAdd(
+                argument,
+                a => a.GetDefaultValue());
         }
 
         internal bool UseDefaultValueFor(IArgument argument)
         {
-            return _argumentsUsingDefaultValue.Contains(argument);
-        }
+            // FIX: (UseDefaultValueFor)
```
Comment (Keboo) [inline src/System.CommandLine/SymbolResult.cs]: I am not sure I understand this comment

File: src/System.CommandLine/SymbolResult.cs [hunk 5]
```diff
@@ -25,293 +28,220 @@ public abstract class SymbolResult
...
+                    return ArgumentResult.Success(argument, symbolResult.Arguments);
+            }
 
-                    default:
-                        return ArgumentResult.Success(argument, symbolResult.Arguments);
-                }
+            bool ShouldCheckArity()
+            {
+                return !(symbolResult is OptionResult optionResult) ||
+                       !optionResult.IsImplicit;
```
Comment (Keboo) [inline src/System.CommandLine/SymbolResult.cs]:
```suggestion
                return symbolResult is OptionResult optionResult && optionResult.IsImplicit;
```

File: src/System.CommandLine/CommandResultExtensions.cs [hunk 6]
```diff
@@ -42,5 +43,39 @@ public static T GetValueOrDefault<T>(this CommandResult commandResult)
...
+        {
+            var children = commandResult.Children
+                                        .Where(o => valueDescriptor.ValueName.IsMatch(o.Symbol))
+                                        .ToArray();
+
+            SymbolResult symbolResult = null;
+
+            if (children.Length > 1)
+            {
+                throw new ArgumentException($"Ambiguous match while trying to bind parameter {valueDescriptor.ValueName} among: {String.Join(",", children.Select(o => o.Name))}");
```
[the remark this answers is not in the document]
Comment (jonsequitur) [reply src/System.CommandLine/CommandResultExtensions.cs]: This will be done at some point but localizing exceptions messages is out of scope for the moment.

File: src/System.CommandLine/SymbolResult.cs (hunk 3 above)
Comment (jonsequitur) [reply src/System.CommandLine/SymbolResult.cs]: This is just a reminder to myself. 

To explain what's in progress here, the `ArgumentResult` versus `ArgumentResult2` issue still has to be sorted out and I'm wondering whether the former needs to remain anywhere in the public API. What I have in mind is something like this:

* `ArgumentResult` and its subclasses get renamed (`ArgumentConversionResult`?) and made internal. Properties that expose it get hidden since the main APIs for retrieving argument results go through `Command` or `Option` anyway (in order to deal gracefully with default or missing values).
* `ArgumentResult2` gets renamed to `ArgumentResult`.

Thoughts?

File: src/System.CommandLine/SymbolResult.cs (hunk 4 above)
Comment (jonsequitur) [reply src/System.CommandLine/SymbolResult.cs]: Removed.

File: src/System.CommandLine/SymbolResult.cs (hunk 3 above)
Comment (Keboo) [reply src/System.CommandLine/SymbolResult.cs]: I like that

File: src/System.CommandLine/SymbolExtensions.cs (hunk 2 above)
Comment (jonsequitur) [reply src/System.CommandLine/SymbolExtensions.cs]: Must be a format setting. This was unintentional.
